lib/rust/parser/src/syntax/operator.rs

```diff
@@ -45,77 +45,80 @@ impl<T> WithPrecedence<T> {
+            ast.with_error("This expression cannot be used in a non-spaced equation."),
+        _ => ast,
+    })
 }
 
-/// Take [`Item`] stream, resolve operators precedence and return the final AST. The precedence
-/// resolution algorithm bases on the [Shunting yard algorithm](https://en.wikipedia.org/wiki/Shunting_yard_algorithm).
-/// It is extended to handle operator sections.
-#[inline(always)]
+/// Take [`Item`] stream, resolve operator precedence and return the final AST.
+///
+/// The precedence resolution algorithm is based on the Shunting yard algorithm[1], extended to
+/// handle operator sections.
+/// [1]: https://en.wikipedia.org/wiki/Shunting_yard_algorithm
 pub fn resolve_operator_precedence<'s>(items: Vec<syntax::Item<'s>>) -> syntax::Tree<'s> {
+    // If `items` is empty, opt_rhs.unwrap() will panic below. Let's panic now with a clear error.
+    assert!(!items.is_empty());
     type Tokens<'s> = Vec<syntax::Item<'s>>;
     let mut flattened: Tokens<'s> = default();
     let mut no_space_group: Tokens<'s> = default();
-    let processs_no_space_group = |flattened: &mut Tokens<'s>, no_space_group: &mut Tokens<'s>| {
-        let tokens = mem::take(no_space_group);
-        if tokens.len() == 1 {
+    let process_no_space_group = |flattened: &mut Tokens<'s>, no_space_group: &mut Tokens<'s>| {
+        let tokens = no_space_group.drain(..);
+        if tokens.len() < 2 {
             flattened.extend(tokens);
         } else {
-            let tokens = annotate_tokens_that_need_spacing(tokens);
+            let tokens = tokens.map(annotate_tokens_that_need_spacing);
             let ast = resolve_operator_precedence_internal(tokens);
             flattened.push(ast.into());
         }
     };
+    let breaks_group = |item: &syntax::item::Item| {
+        if item.left_visible_offset().width_in_spaces != 0 {
+            return true;
+        }
+        if let syntax::item::Item::Block(_) = item {
+            return true;
+        }
+        false
+    };
     for item in items {
-        if item.left_visible_offset().width_in_spaces == 0 || no_space_group.is_empty() {
-            no_space_group.push(item)
-        } else if !no_space_group.is_empty() {
-            processs_no_space_group(&mut flattened, &mut no_space_group);
-            no_space_group.push(item);
-        } else {
-            // FIXME: this is unreachable.
-            flattened.push(item);
+        if breaks_group(&item) {
+            process_no_space_group(&mut flattened, &mut no_space_group);
         }
+        no_space_group.push(item);
     }
-    if !no_space_group.is_empty() {
-        processs_no_space_group(&mut flattened, &mut no_space_group);
-    }
+    process_no_space_group(&mut flattened, &mut no_space_group);
     resolve_operator_precedence_internal(flattened)
 }
 
-fn resolve_operator_precedence_internal(items: Vec<syntax::Item<'_>>) -> syntax::Tree<'_> {
+#[derive(PartialEq, Eq)]
+enum TokenType {
+    Ast,
+    Opr,
+}
```

Comment:
1. This enum is kind of strange. We have `Token` defined and tokens cannot be AST.
2. Tokens has a lot of different types, none of them is AST.
3. This is defined in the "Precedence" section - is it a precedence-related definition?
4. It should have docs, as its usage patterns are not clear by looking at its name.

Reply:
This replaces `last_token_was_ast`/`last_token_was_opr`. I have fixed the name--it actually classifies an `Item`, not a token. I have documented it and moved it into `resolve_operator_precedence_internal`, as it is only used to keep track of state in that function.